Add security docs (#152)
canonical/dns-operators: docs/explanation/security.md

# Security

This document outlines common risks and possible best practices specifically for the DNS charms. It
focuses on configurations and protections available through the charms themselves.

The overall best practice is to [keep your charms updated](https://documentation.ubuntu.com/juju/3.6/reference/juju-cli/list-of-juju-cli-commands/refresh/) to the latest version available.  
A good understanding of [the DNS system](https://bind9.readthedocs.io/en/stable/chapter1.html) is welcome.

## Configuration

The DNS charms can be configured and misconfiguring them can lead to security vulnerabilities.  

### bind-operator

> No configuration available.

Reviewer: I prefer the description you used for `dns-integrator`:
```suggestion
No security-related configurations available.
```

### dns-policy

> `dns-policy` is using a Django application under the hood for its API and web interface. As such, it can be configured following

Reviewer: ```suggestion
`dns-policy` uses a Django application under the hood for its API and web interface. As such, it can be configured following
```

> `dns-policy` uses a Django application under the hood for its API and web interface. As such, it can be configured following

Reviewer 2: What about adding advice on how to store the user/password for accessing the application, and what could happen if someone with bad intentions gains access to it?

Author: Added a paragraph to explain a bit more about reviewers and the security implications of these users.

> `dns-policy` uses a Django application under the hood for its API and web interface. As such, it can be configured following

Reviewer 2: Also, any recommendations if the database is lost?

some of Django's configurations.  
- **debug**: Puts the application in [debug mode](https://docs.djangoproject.com/en/stable/ref/settings/#std-setting-DEBUG).
- **allowed-hosts**: Configures the [hosts allowed](https://docs.djangoproject.com/en/stable/ref/settings/#std-setting-ALLOWED_HOSTS) to reach the API and web interface of the application.

### dns-integrator

> No security related configuration available

Reviewer: ```suggestion
No security-related configurations available.
```
I would use a hyphen here to clarify that it's a single adjective

## Data

> Only the dns-policy is using an external database to store its data, through a postgresql interface.

Reviewer: ```suggestion
Only `dns-policy` is using an external database to store its data, through a `postgresql` interface.
```
I think `postgresql` should be code-formatted to avoid complaints from the style checker

### Back up PostgreSQL

Follow the instructions of the PostgreSQL charm:

>  - For [postgresql-k8s](https://charmhub.io/postgresql-k8s/docs/h-create-backup).
>  - For [postgresql](https://charmhub.io/postgresql/docs/h-create-backup).

Reviewer: ```suggestion
 - For [`postgresql-k8s`](https://charmhub.io/postgresql-k8s/docs/h-create-backup).
 - For [`postgresql`](https://charmhub.io/postgresql/docs/h-create-backup).
```

If you plan to restore PostgreSQL in a different model or cluster, you will need
to also back up the cluster passwords. See:

>  - For [postgresql-k8s](https://charmhub.io/postgresql-k8s/docs/h-migrate-cluster).
>  - For [postgresql](https://charmhub.io/postgresql/docs/h-migrate-cluster).

Reviewer: ```suggestion
 - For [`postgresql-k8s`](https://charmhub.io/postgresql-k8s/docs/h-migrate-cluster).
 - For [`postgresql`](https://charmhub.io/postgresql/docs/h-migrate-cluster).
```

### Restore PostgreSQL

Follow the instructions given by PostgreSQL:

>  - For postgresql-k8s: [local restore](https://charmhub.io/postgresql/docs/h-restore-backup), [foreign backup](https://charmhub.io/postgresql/docs/h-migrate-cluster).
>  - for postgresql: [local restore](https://charmhub.io/postgresql/docs/h-restore-backup), [foreign backup](https://charmhub.io/postgresql/docs/h-migrate-cluster).

Reviewer: ```suggestion
 - For `postgresql-k8s`: [local restore](https://charmhub.io/postgresql/docs/h-restore-backup), [foreign backup](https://charmhub.io/postgresql/docs/h-migrate-cluster).
 - For `postgresql`: [local restore](https://charmhub.io/postgresql/docs/h-restore-backup), [foreign backup](https://charmhub.io/postgresql/docs/h-migrate-cluster).
```
